pkg/services/alerting/notifiers/slack.go

```diff
@@ -346,37 +346,47 @@ func slackFileUpload(evalContext *alerting.EvalContext, log log.Logger, url stri
 	return nil
 }
 
-func generateSlackBody(file string, token string, recipient string) (map[string]string, bytes.Buffer, error) {
+func (sn *SlackNotifier) generateSlackBody(path string, token string, recipient string) (map[string]string, bytes.Buffer, error) {
 	// Slack requires all POSTs to files.upload to present
 	// an "application/x-www-form-urlencoded" encoded querystring
 	// See https://api.slack.com/methods/files.upload
 	var b bytes.Buffer
 	w := multipart.NewWriter(&b)
+	defer func() {
+		if err := w.Close(); err != nil {
+			// Shouldn't matter since we already close w explicitly on the non-error path
+			sn.log.Warn("Failed to close multipart writer", "err", err)
+		}
+	}()
+
 	// Add the generated image file
-	f, err := os.Open(file)
+	f, err := os.Open(path)
 	if err != nil {
 		return nil, b, err
 	}
-	defer f.Close()
-	fw, err := w.CreateFormFile("file", file)
+	defer func() {
+		if err := f.Close(); err != nil {
+			sn.log.Warn("Failed to close file", "path", path, "err", err)
+		}
+	}()
+	fw, err := w.CreateFormFile("file", path)
 	if err != nil {
 		return nil, b, err
 	}
-	_, err = io.Copy(fw, f)
-	if err != nil {
+	if _, err := io.Copy(fw, f); err != nil {
 		return nil, b, err
 	}
 	// Add the authorization token
-	err = w.WriteField("token", token)
-	if err != nil {
+	if err := w.WriteField("token", token); err != nil {
 		return nil, b, err
 	}
 	// Add the channel(s) to POST to
-	err = w.WriteField("channels", recipient)
-	if err != nil {
+	if err := w.WriteField("channels", recipient); err != nil {
 		return nil, b, err
 	}
-	w.Close()
+	if err := w.Close(); err != nil {
```

Comment:
Any reason for closing it twice? Shouldn't the deferred call on the top be good enough?

Reply:
@bergquist Yes there is a reason :) It's actually a known pattern in Go, because returning an error from `defer` can be tricky. Either you have to change the function to use named errors, so `defer` can return an error, or you do as I've done here, i.e. close explicitly and return any error, and close in a `defer` just so you make sure it's closed also on error paths.

The second close shouldn't matter from what I've read, it's safe to close a resource several times.